contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,207 @@
- */
-contract TokenPaymaster is BasePaymaster, ERC20 {
-
-    //calculated cost of the postOp
-    uint256 constant public COST_OF_POST = 15000;
-
-    address public immutable theFactory;
-
-    constructor(address accountFactory, string memory _symbol, IEntryPoint _entryPoint) ERC20(_symbol, _symbol) BasePaymaster(_entryPoint) {
-        theFactory = accountFactory;
-        //make it non-empty
-        _mint(address(this), 1);
-
-        //owner is allowed to withdraw tokens from the paymaster's balance
-        _approve(address(this), msg.sender, type(uint).max);
+    event ConfigUpdated(TokenPaymasterConfig tokenPaymasterConfig);
+
+    event UserOperationSponsored(address indexed user, uint256 actualTokenCharge, uint256 actualGasCost, uint256 actualTokenPrice);
+
+    event PostOpReverted(address indexed user, uint256 preCharge);
+
+    event Received(address indexed sender, uint256 value);
+
+    /// @notice All 'price' variables are multiplied by this value to avoid rounding up
+    uint256 private constant PRICE_DENOMINATOR = 1e6;
+
+    /// @notice Estimated gas cost for refunding tokens after the transaction is completed
+    uint256 public constant REFUND_POSTOP_COST = 40000;
+
+    TokenPaymasterConfig private tokenPaymasterConfig;
+
+    // TODO: I don't like defaults in Solidity - accept ALL parameters of fail!!!
+    /// @notice Initializes the PimlicoERC20Paymaster contract with the given parameters.
+    /// @param _token The ERC20 token used for transaction fee payments.
+    /// @param _entryPoint The EntryPoint contract used in the Account Abstraction infrastructure.
+    /// @ param _tokenOracle The Oracle contract used to fetch the latest token prices.
+    /// @ param _nativeAssetOracle The Oracle contract used to fetch the latest native asset (ETH, Matic, Avax, etc.) prices.
+    /// @param _owner The address that will be set as the owner of the contract.
+    constructor(
+        IERC20Metadata _token,
+        IEntryPoint _entryPoint,
+        IERC20 _wrappedNative,
+        ISwapRouter _uniswap,
+        TokenPaymasterConfig memory _tokenPaymasterConfig,
+        OracleHelperConfig memory _oracleHelperConfig,
+        UniswapHelperConfig memory _uniswapHelperConfig,
+        address _owner
+    )
+    BasePaymaster(
+    _entryPoint
+    )
+    OracleHelper(
+    _oracleHelperConfig,
+    10 ** _token.decimals()
+    )
+    UniswapHelper(
+    _token,
+    _wrappedNative,
+    _uniswap,
+    _uniswapHelperConfig
+    )
+    {
+        setTokenPaymasterConfig(_tokenPaymasterConfig);
+        transferOwnership(_owner);
     }
 
+    /// @notice Updates the configuration for the Token Paymaster.
+    /// @param _tokenPaymasterConfig The new price markup percentage (1e6 = 100%).
+    function setTokenPaymasterConfig(
+        TokenPaymasterConfig memory _tokenPaymasterConfig
+    ) public onlyOwner {
+        require(_tokenPaymasterConfig.priceMarkup <= 1900000, "TPM: price markup too high");
```

Comment:
Out of curiosity, how do we pick the upper bound (1900000)?

Reply:
None of these hard-coded parameters are not yet selected, but in my opinion the maximum possible value for the markup should just be there to prevent human error or malicious configuration.

For example, Pimlico Paymaster used `120e4` which sounds to me like it may be a bit too limiting:
https://github.com/pimlicolabs/erc20-paymaster-contracts/blob/c9911402a49928398441b07fb9caa9f4b4d0352f/src/PimlicoERC20Paymaster.sol#L68

Reply:
maybe 2e6 is good upper limit

Reply:
@livingrockrises Agreed. Will set for 200% for now. Will also make it explicitly related to the `PRICE_DENOMINATOR` value.